### PR TITLE
feat: populate the index on snapshot import

### PR DESCRIPTION
**Pull request by fridrik01**
## Related Issues
Fixes: https://github.com/filecoin-project/lotus/issues/10537

## Proposed Changes
After importing from snapshot this PR creates and populates the sqlite message index introduced in https://github.com/filecoin-project/lotus/pull/10452 from the 2k tipsets in the snapshot.

## Additional Info
When loading from snapshot from mainnet, populating the message index takex approx 30 seconds. We should be able to lower this somewhat with bulk inserts (though doing it inside a transaction is at least an order of magnitude faster than not).

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/index/msgindex.go [hunk 1]
```diff
@@ -169,6 +169,68 @@ func NewMsgIndex(lctx context.Context, basePath string, cs ChainStore) (MsgIndex
...
+		}
+
+		tskey := tscid.String()
+		epoch := int64(curTs.Height())
+
+		//log.Infof("epoch %d-%d, populating msgindex with tipset %s", curTs.Height(), startHeight-curTs.Height(), tskey)
+
+		msgs, err := cs.MessagesForTipset(lctx, curTs)
+		if err != nil {
+			log.Infof("stopping import after %d tipsets", startHeight-curTs.Height())
```
Comment (fridrik01) [inline chain/index/msgindex.go:214]: @vyzo is there a better way to break from the loop other than doing this hack (or hard-coding 2000 tipsets)?

Comment (vyzo) [reply chain/index/msgindex.go:214]: i dont think so... maybe have height condition instead of niling curTs?

File: chain/index/msgindex.go [hunk 2]
```diff
@@ -169,6 +169,85 @@ func NewMsgIndex(lctx context.Context, basePath string, cs ChainStore) (MsgIndex
...
 
+func PopulateAfterSnapshot(lctx context.Context, basePath string, cs ChainStore) error {
+	err := os.MkdirAll(basePath, 0755)
+	if err != nil {
+		return xerrors.Errorf("error creating msgindex base directory: %w", err)
+	}
+
+	dbPath := path.Join(basePath, dbName)
+	if _, err := os.Stat(dbPath); err == nil {
+		return xerrors.Errorf("msgindex already exists at %s", dbPath)
```
Comment (vyzo) [inline chain/index/msgindex.go:180]: maybe just delete it and recreate it?
Sounds like the right thing to do.

**Review by vyzo — COMMENTED** [review]
overall lgtm, just a small comment.

**Comment by fridrik01** [comment]
@vyzo I double checked and verified that the lotus daemon continues to populate and update the `sqlite/msgindex.db` created by populating the message index after loading from snapshot, so looking good.

File: chain/index/msgindex.go [hunk 3]
```diff
@@ -169,6 +169,97 @@ func NewMsgIndex(lctx context.Context, basePath string, cs ChainStore) (MsgIndex
...
+
+	rollback := func() {
+		if err := tx.Rollback(); err != nil {
+			log.Errorf("error in rollback: %s", err)
+		}
+	}
+
+	insertStmt, err := tx.Prepare(dbqInsertMessage)
+	if err != nil {
+		return xerrors.Errorf("error preparing insertStmt: %w", err)
```
Comment (vyzo) [inline chain/index/msgindex.go:214]: rollback

File: chain/index/msgindex.go [hunk 4]
```diff
@@ -169,6 +169,97 @@ func NewMsgIndex(lctx context.Context, basePath string, cs ChainStore) (MsgIndex
...
+	}
+	defer func() {
```
Comment (vyzo) [inline chain/index/msgindex.go:216]: unnecessary, the transaction is closed.

**Review by vyzo — COMMENTED** [review]
a couple of small things.

**Review by vyzo — APPROVED** [review]

**Review by arajasek — APPROVED** [review]

**Pull request merged**